### ISSUE TITLE
mIoU evaluation for VOC2012 and COCO17

### ISSUE THREAD

**Issue by SamyakR99** [xmed-lab/CLIP_Surgery#32]
Thanks for a great work!
@Eli-YiLi 

I have several questions regarding the mIoU evaluation for VOC12 and COCO17 using ResNet 101:

1. Could you please describe your procedure for obtaining mIoU scores after obtaining the similarity maps?
2. I read in a previous issue [(#17 )](https://github.com/xmed-lab/CLIP_Surgery/issues/17) that for open vocab segmentation, the segmentation map is obtained by simply applying argmax to the similarity maps. 
3. I used the segmentation maps from argmax and the gt masks to get the mIoU. 

The results I obtained were: VOC2012: 38.90 and COCO17: 0.11. 

I am surely missing something here as the mIoU for COCO is significantly lower. I would really appreciate it if you could help me on this.

Thanks!

**Comment by Eli-YiLi** [comment]
Which task you are doing?

For explainability task follow this please:
![image](https://github.com/user-attachments/assets/f8bcf6af-c151-4595-ba5e-37647f48c902)
![image](https://github.com/user-attachments/assets/22386add-34c9-40a8-8db6-ce554057d529)

For segmentation task, just apply argmax on foreground classes, and count IoU of each positive classes, then get the mIoU.

**Comment by SamyakR99** [comment]
@Eli-YiLi 

Thanks for the prompt reply!

I am working on the segmentation task, specifically replicating the mIoU values in the Table 3 of the paper
![image](https://github.com/user-attachments/assets/7799a857-6ead-477f-8c61-3b520fd1a0d3)

I have a follow-up question:  
 Could you clarify what positive classes mean? and why compute it over only the positive ones. Should we not compute it over all the classes? 

'''
    I assumed **positive classes are the ones after argmax.** 
    
     Say  
    pred: 1,2,3
    GT: 1,2,3,4

    if you only do the positive classes then you only get performance based on IoU of the 1st, 2nd and 3rd class. However, the 
    IoU value for the 4th is 0, which is not considered and itwill end up reducing the mIoU. 


    The same argument can be made if we predict more classes 
    pred: 1,2,3 
    GT: 1,2

    In this case, the mIoU is correct as IoU will be 0 for the 3rd class and we considered it in our calculation. 


    I think the best case that covers the both is to go over the union of clases of pred and GT or simply go over all the classes and ignore the nan values that appear due to absence in both pred and GT.  
'''


I tried iterating over all the classes  on VOC2012  with RN101 and the performance I obtained is having high deviation from Table 3.   

I am attaching my code for VOC2012 dataset. Let me know if you also need to look at my dataloader in case you run it at your side.

'''
    import torch
    from tqdm import tqdm

    def calculate_iou(pred, target, num_classes=21):
        ious = []
        pred = pred.view(-1)
        target = target.view(-1)

        for cls in range(1, num_classes): ## start from 1 to ignore background
            pred_inds = pred == cls
            target_inds = target == cls
            intersection = torch.sum(pred_inds[target_inds])
            union = torch.sum(pred_inds) + torch.sum(target_inds) - intersection

            if union == 0:
                ious.append(float('nan'))
            else:
                ious.append(float(intersection) / max(union, 1))
    #     print(ious)
    #     print(np.nanmean(ious))
        return np.nanmean(ious)


    miou_list = []

    with torch.no_grad():
        i = 0
        iou_scores = []
        for images, cv2_img, targets, pre_target in tqdm(test_loader):

            images = images.to(device)
            targets = targets.to(device)

            image_features = model.encode_image(images)

            features = image_features @ text_features.t()
            similarity = clip.clip_feature_surgery(image_features, text_features)
            similarity_map = clip.get_similarity_map(similarity[:, 1:, :], 224)

            pred = similarity_map.argmax(dim = -1)
            pred = pred + 1 ## making sure that 0 is background

            targets = 255.*targets
            targets[targets == 255] = 0 ## only keep 0 for background, 255 was border, made it background

            target_mask_list = []
            pred_mask_list = []

            for val in range(1,21):
                target_mask_list.append((targets.squeeze(0) == val).to(int)*(val))
                pred_mask_list.append((pred == val).to(int)*val)

            target_mask = sum(target_mask_list)
            pred_mask = sum(pred_mask_list)

            ans = calculate_iou(pred_mask.cpu(), target_mask.cpu())
            miou_list.append(ans)

    print(np.nanmean(miou_list))
    
'''

**Comment by Eli-YiLi** [comment]
Tab. 3 is explainability task, and segmentation task is shown in tab.6 using PASCAL Context COCO Stuf instead.

**Comment by SamyakR99** [comment]
Oh.. gotcha! Thanks

**Comment by SamyakR99** [comment]
@Eli-YiLi 

Hi,

I am having trouble finding the PASCAL context dataset. The one commonly used by segmentation approaches (about 5k images). 
I saw on their website that we should download the test set of  VOC2010 from official website. However, that has a lot more images (more than 5k used in the test) and no ground truth for context dataset. I would really appreciate it if you could point me  to links of the datasets.

Thanks!
